cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -85,22 +129,52 @@ func RetrieveValidatedClusterInfo(discoveryToken string, tokenAPIServers []strin
 			return nil, fmt.Errorf("couldn't parse the kubeconfig file in the %s configmap: %v", bootstrapapi.ConfigMapClusterInfo, err)
 		}
 
-		fmt.Printf("[discovery] Cluster info signature and contents are valid, will use API Server %q\n", bootstrapConfig.Clusters[clusterName].Server)
+		// If no TLS root CA pinning was specified, we're done.
+		if pubKeyPins.Empty() {
+			fmt.Printf("[discovery] Cluster info signature and contents are valid and no TLS pinning was specified, will use API Server %q\n", endpoint)
+			return finalConfig, nil
+		}
+
+		// For each cluster, validate that its CA matches a pinned key and save that root into an x509.CertPool
+		pinnedRoots := x509.NewCertPool()
+		for _, cluster := range finalConfig.Clusters {
+			caCert, err := getClusterCA(cluster)
+			if err != nil {
+				return nil, fmt.Errorf("could not get cluster CA from %s configmap: %v", bootstrapapi.ConfigMapClusterInfo, err)
+			}
+
+			err = pubKeyPins.Check(caCert)
+			if err != nil {
+				return nil, fmt.Errorf("unknown cluster CA: %v", err)
+			}
+
+			pinnedRoots.AddCert(caCert)
+		}
+
+		// Build an intermediate CA pool containing all the non-leaf certificates sent by the server
+		chainCerts := x509.NewCertPool()
+		for _, chainCert := range bootstrapCertificateChain[1:] {
+			chainCerts.AddCert(chainCert)
+		}
+
+		// Now that we know the root CA pool, validate the original certificate chain
```

Comment:
Making double requests seemed worse to me, but I think it's sound from a security perspective.